CenterForOpenScience/osf.io
# [PLAT-462] Project Access Requests Frontend

## Purpose

Allow users to request access to a public or a private project that they are not currently contributors on. Project admins can allow/disallow requests for access on the project's settings page, and approve/deny existing requests on the project's contributors page.

**NOTE**: also includes @chennan47's changes to [add node logs when project access is updated](https://github.com/erinspace/osf.io/pull/16)


## Changes

- Ability for admins to enable/disable project access requests on the project settings page
- Custom template for 403 page when accessing a private project that has access requests enabled
- Button on public projects that allow requests to be made by non-contributors with access requests enabled
- Admins can see a list of access requests on the contributor page and choose to reject, or accept them while giving certain permissions/setting visible
- Change UI on project page to move the share/bookmark buttons under a [...] dropdown menu
- Migration/model change to request model to allow for partial unique index ignoring when the machine state is `accepted`
- Migration/model change to `NodeRequestAction` to allow passing permissions and contributor visibility levels when the request is accepted from the contributors page
- Change text email templates to html, for consistency with #8187
- Change red [Remove] buttons for Contributors and VOLs to Xs
- Those Xs should become grey default "Remove" buttons when screen is small enough for "card mode"

## QA Notes



This will need a lot of QA and has many edge cases. I'll try to be as thorough as possible here but I could totally miss something that you guys' expertise will catch!

[Check out the parent ticket](https://openscience.atlassian.net/browse/PLAT-473) for requirements and the [prototype mockups](https://marvelapp.com/b88hg84) for more info.

1. Ability to request access
- Access Requests Enabled
    - A non-contributor user visits a private project with access requests enabled (enabled by default) and sees a special 403 page with an option to request access
    - A non-contributor user visits a public project with access requests enabled (enabled by default) and sees an option to request access under the new [...] button as a dropdown menu
    - A contributor visiting a project should not see the request access button on the [...] menu
- Access Requests Disabled
    - An  admin disables request access on a private project on the project settings page
    - An admin disables request access on a  public project on the project settings page
    - A non-contributor visits the private project, gets the standard usual "user has restricted access" 403 page
    - a non-contributor visits the public project, does not see the 'request access' option on the [...] dropdown menu

2. Requesting Access
- A user requests access to a project (pretty much the same should happen for public and private)
    - Clicking on the button/link immediately disables it
    - After a request is completed, the button/link should still be disabled but now read "access requested"
    - Project admins get an email that a user has requested access to their project
- Private projects with request access enabled also have a "switch account to request access" button 
    - clicking this takes them to CAS
    - After logging in with another account through cas they are taken back to the project as a contributor if they have permission, they see the same request access screen if not

- A user's access request is rejected
    - Visiting the project, the disabled "access requested" link/button now has a tooltip that says "access denied"
    - The requesting user gets an email that their request was denied

- A user's access request is accepted
    - The user can now visit the project/be included as a contributor
    - The user gets an email that their request for access has been accepted

- A logged out user visiting a public project with access requests enabled
    - The user sees "login to request access" in the [...] dropdown menu
    - Clicking this button takes them to CAS to login
    - After loging in through CAS, they are returned to the project page

3. Accepting/Rejecting access requests
- An admin should see a list of Requests for access on the contributors page
    - This list should behave much the same on a small screen as the contributors list, collapse into a card form on super small screens
    - Each request has the option to reject (X) or accept (+Add)
    - Rejecting a request removes the request from the list and emails the requesting user
    - Accepting a request adds the user as a contributor with the permissions/visibility selected in the row when the 'add' button is clicked
        - the page should be refreshed, the user appears in the contributor list with the proper permissions, and the request is removed from the list
- A non-admin should not see the Requests for Access Section
- When there are no requests, even an admin should not see the Requests for Access section

Edge cases to check:
- A user requests access to a project, is accepted as a contributor, is removed as a contributor, and requests access again
- Components should be separate of their parents. If someone requests access to a toplevel and gets rejected, they should still be able to independently request the child component
- Probably others! Ah!

Unrelated-ish things to check:
- Contributors page UI changes: Formerly red [Remove] button is now a gray X
- That gray X becomes a default [Remove] button when the screen size is small
- The facebook sharing link was acting a little funny earlier, but might be unrelated to this PR -- in any case, might be good to check that it's all a-ok

## Side Effects

None anticipated as this is a pretty new feature

## Ticket
https://openscience.atlassian.net/browse/PLAT-462
